File: phases/16-software-engineering-and-architecture/18-observability-as-a-design-concern/code/main.ts
interface LogEntry {
  timestamp: string;
  level: "TRACE" | "DEBUG" | "INFO" | "WARN" | "ERROR" | "FATAL";
  message: string;
  service: string;
  trace_id?: string;
  span_id?: string;
  route?: string;
  status_code?: number;
  latency_ms?: number;
  error?: string;
  user_id?: string;
}

interface Span {
  traceId: string;
  spanId: string;
  parentId?: string;
  operation: string;
  startTime: number;
  endTime?: number;
  attributes: Record<string, string>;
  status: "ok" | "error";
}

interface MetricPoint {
  name: string;
  value: number;
  labels: Record<string, string>;
  timestamp: number;
  type: "counter" | "gauge" | "histogram";
}

const SERVICE_NAME = "api-server";
const spanStorage: Span[] = [];
const metricStorage: MetricPoint[] = [];
const histogramBuckets = [5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000];
const latencyHistogram: Record<string, number> = {};
let latencySum = 0;
let latencyCount = 0;

for (const bound of histogramBuckets) {
  latencyHistogram[`le_${bound}`] = 0;
}
latencyHistogram["le_inf"] = 0;

function emitLog(entry: Partial<LogEntry> & { message: string }): void {
  const full: LogEntry = {
    timestamp: new Date().toISOString(),
    level: entry.level ?? "INFO",
    message: entry.message,
    service: entry.service ?? SERVICE_NAME,
    ...entry,
  } as LogEntry;
  console.log(JSON.stringify(full));
}

function generateId(length: number): string {
  const chars = "0123456789abcdef";
  let id = "";
  for (let i = 0; i < length; i++) {
    id += chars[Math.floor(Math.random() * chars.length)];
  }
  return id;
}

function generateTraceId(): string {
  return generateId(16);
}

function generateSpanId(): string {
  return generateId(8);
}

interface RequestContext {
  traceId: string;
  spanId: string;
  parentId?: string;
}

function createRequestContext(
  incomingTraceId?: string,
  parentSpanId?: string
): RequestContext {
  const traceId = incomingTraceId ?? generateTraceId();
  const spanId = generateSpanId();
  return { traceId, spanId, parentId: parentSpanId };
}

function startSpan(
  ctx: RequestContext,
  operation: string,
  attributes?: Record<string, string>
): Span {
  const span: Span = {
    traceId: ctx.traceId,
    spanId: ctx.spanId,
    parentId: ctx.parentId,
    operation,
    startTime: Date.now(),
    attributes: attributes ?? {},
    status: "ok",
  };
  spanStorage.push(span);
  return span;
}

function endSpan(span: Span): void {
  span.endTime = Date.now();
}

function recordCounter(
  name: string,
  value: number,
  labels: Record<string, string>
): void {
  metricStorage.push({
    name,
    value,
    labels,
    timestamp: Date.now(),
    type: "counter",
  });
}

function recordLatency(ms: number): void {
  latencySum += ms;
  latencyCount += 1;
  for (const bound of histogramBuckets) {
    if (ms <= bound) {
      latencyHistogram[`le_${bound}`] += 1;
    }
  }
  latencyHistogram["le_inf"] += 1;
}

function recordRedMetrics(
  route: string,
  statusCode: number,
  latencyMs: number
): void {
  recordCounter("http_requests_total", 1, { route, method: "GET" });
  if (statusCode >= 400) {
    recordCounter("http_errors_total", 1, {
      route,
      method: "GET",
      status_code: String(statusCode),
    });
  }
  recordLatency(latencyMs);
}

function formatMetrics(): string {
  const lines: string[] = [];
  const agg: Record<string, { value: number; type: string; labels: Record<string, string> }> = {};
  for (const m of metricStorage) {
    const key = `${m.name}:${JSON.stringify(m.labels)}`;
    if (!agg[key]) {
      agg[key] = { value: 0, type: m.type, labels: m.labels };
    }
    if (m.type === "counter") {
      agg[key].value += m.value;
    } else if (m.type === "gauge") {
      agg[key].value = m.value;
    }
  }
  lines.push("# HELP http_requests_total Total HTTP requests");
  lines.push("# TYPE http_requests_total counter");
  lines.push(`http_requests_total ${Object.values(agg).find(a => Object.keys(a.labels).includes("route") && a.labels.route)?.value ?? 0}`);

  lines.push("# HELP http_errors_total Total HTTP errors");
  lines.push("# TYPE http_errors_total counter");

  lines.push("# HELP http_request_duration_ms Histogram of request durations");
  lines.push("# TYPE http_request_duration_ms histogram");
  let cumulative = 0;
  for (const bound of histogramBuckets) {
    cumulative += latencyHistogram[`le_${bound}`];
    lines.push(`http_request_duration_ms_bucket{le="${bound}"} ${cumulative}`);
  }
  lines.push(`http_request_duration_ms_bucket{le="+Inf"} ${latencyHistogram["le_inf"]}`);
  lines.push(`http_request_duration_ms_sum ${latencySum}`);
  lines.push(`http_request_duration_ms_count ${latencyCount}`);
  return lines.join("\n");
}

interface HandlerResult {
  statusCode: number;
  body: Record<string, unknown>;
  error?: string;
}

function healthHandler(_ctx: RequestContext): HandlerResult {
  return { statusCode: 200, body: { status: "ok" } };
}

function usersHandler(ctx: RequestContext): HandlerResult {
  const span = startSpan(ctx, "GET /users", { "http.method": "GET" });
  const userId = String(Math.floor(Math.random() * 1000) + 1);

  const childCtx: RequestContext = {
    traceId: ctx.traceId,
    spanId: generateSpanId(),
    parentId: ctx.spanId,
  };

  startSpan(childCtx, "db.query", {
    "db.system": "postgres",
    "db.operation": "SELECT",
    "user.id": userId,
  });

  emitLog({
    level: "DEBUG",
    message: "fetching user from database",
    service: "user-service",
    trace_id: ctx.traceId,
    span_id: childCtx.spanId,
    route: "GET /users",
    user_id: userId,
  });

  if (Math.random() < 0.15) {
    const delayMs = Math.floor(Math.random() * 2000) + 500;
    span.status = "error";
    span.attributes["error.type"] = "timeout";
    endSpan(span);
    return {
      statusCode: 504,
      body: { error: "database timeout" },
      error: `timeout after ${delayMs}ms querying user ${userId}`,
    };
  }

  endSpan(span);
  return {
    statusCode: 200,
    body: { id: userId, name: `user_${userId}`, email: `${userId}@example.com` },
  };
}

function ordersHandler(ctx: RequestContext): HandlerResult {
  const span = startSpan(ctx, "GET /orders", { "http.method": "GET" });
  const orderId = String(Math.floor(Math.random() * 5000) + 1);

  const childCtx: RequestContext = {
    traceId: ctx.traceId,
    spanId: generateSpanId(),
    parentId: ctx.spanId,
  };

  startSpan(childCtx, "http.client", {
    "http.method": "POST",
    "http.url": "http://payment-service/charge",
    "order.id": orderId,
  });

  emitLog({
    level: "DEBUG",
    message: "calling payment service for order",
    service: "order-service",
    trace_id: ctx.traceId,
    span_id: childCtx.spanId,
    route: "GET /orders",
    user_id: orderId,
  });

  if (Math.random() < 0.1) {
    span.status = "error";
    span.attributes["error.type"] = "upstream_error";
    endSpan(span);
    return {
      statusCode: 502,
      body: { error: "payment service unavailable" },
      error: `payment service error for order ${orderId}`,
    };
  }

  endSpan(span);
  return {
    statusCode: 200,
    body: { order_id: orderId, status: "confirmed", total: "29.99" },
  };
}

type RouteHandler = (ctx: RequestContext) => HandlerResult;

function observeHandler(
  handler: RouteHandler,
  route: string
): (traceparent?: string) => HandlerResult {
  return (traceparent?: string): HandlerResult => {
    const incomingTraceId = traceparent ?? undefined;
    const ctx = createRequestContext(incomingTraceId);
    const start = Date.now();

    const result = handler(ctx);

    const latencyMs = Date.now() - start;
    recordRedMetrics(route, result.statusCode, latencyMs);

    const level = result.statusCode >= 500 ? "ERROR" : result.statusCode >= 400 ? "WARN" : "INFO";
    emitLog({
      level: level as LogEntry["level"],
      message: "request completed",
      trace_id: ctx.traceId,
      span_id: ctx.spanId,
      route,
      status_code: result.statusCode,
      latency_ms: latencyMs,
      error: result.error,
    });

    return result;
  };
}

function simulateTraffic(): void {
  const routes: { path: string; handler: RouteHandler }[] = [
    { path: "/health", handler: healthHandler },
    { path: "/users", handler: usersHandler },
    { path: "/orders", handler: ordersHandler },
  ];

  const observed = routes.map((r) => ({
    path: r.path,
    handler: observeHandler(r.handler, `GET ${r.path}`),
  }));

  console.log("=== Simulating observed HTTP traffic ===\n");

  for (let i = 0; i < 20; i++) {
    const route = observed[Math.floor(Math.random() * observed.length)];
    const traceId = i < 3 ? "parent-trace-abc123" : undefined;
    const result = route.handler(traceId);
    console.log(
      `  -> ${route.path} => ${result.statusCode} ${JSON.stringify(result.body).slice(0, 60)}`
    );
  }

  console.log("\n=== Distributed Traces ===\n");
  const traces: Record<string, Span[]> = {};
  for (const span of spanStorage) {
    if (!traces[span.traceId]) traces[span.traceId] = [];
    traces[span.traceId].push(span);
  }

  for (const [traceId, spans] of Object.entries(traces).slice(0, 5)) {
    console.log(`Trace: ${traceId}`);
    for (const span of spans) {
      const duration = span.endTime ? span.endTime - span.startTime : "?";
      const parent = span.parentId ? ` (parent: ${span.parentId})` : "";
      console.log(
        `  Span ${span.spanId}: ${span.operation} [${duration}ms]${parent} status=${span.status}`
      );
    }
    console.log("");
  }

  console.log("=== Prometheus Metrics ===\n");
  console.log(formatMetrics());

  console.log("\n=== Observability Summary ===");
  console.log(`Total requests: ${latencyCount}`);
  console.log(`Total errors: ${metricStorage.filter(m => m.name === "http_errors_total").length}`);
  console.log(`Avg latency: ${latencyCount > 0 ? Math.round(latencySum / latencyCount) : 0}ms`);
  console.log(`Total spans: ${spanStorage.length}`);
  console.log(`Total traces: ${Object.keys(traces).length}`);
}

simulateTraffic();

File: phases/16-software-engineering-and-architecture/18-observability-as-a-design-concern/code/main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"os"
	"strconv"
	"sync/atomic"
	"time"
)

type LogEntry struct {
	Timestamp  string `json:"timestamp"`
	Level      string `json:"level"`
	Message    string `json:"message"`
	Service    string `json:"service"`
	TraceID    string `json:"trace_id,omitempty"`
	SpanID     string `json:"span_id,omitempty"`
	Route      string `json:"route,omitempty"`
	StatusCode int    `json:"status_code,omitempty"`
	LatencyMs  int64  `json:"latency_ms,omitempty"`
	Error      string `json:"error,omitempty"`
	UserID     string `json:"user_id,omitempty"`
}

var (
	logger = log.New(os.Stdout, "", 0)

	requestTotal   atomic.Int64
	errorTotal     atomic.Int64
	latencySum     atomic.Int64
	latencyCount   atomic.Int64
	latencyBuckets [12]atomic.Int64

	bucketBounds = []float64{5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000}
)

func emitLog(entry LogEntry) {
	if entry.Timestamp == "" {
		entry.Timestamp = time.Now().UTC().Format(time.RFC3339Nano)
	}
	data, _ := json.Marshal(entry)
	logger.Println(string(data))
}

func generateTraceID() string {
	b := make([]byte, 16)
	for i := range b {
		b[i] = "0123456789abcdef"[rand.Intn(16)]
	}
	return string(b)
}

func generateSpanID() string {
	b := make([]byte, 8)
	for i := range b {
		b[i] = "0123456789abcdef"[rand.Intn(16)]
	}
	return string(b)
}

type contextKey string

const (
	traceIDKey contextKey = "trace_id"
	spanIDKey  contextKey = "span_id"
)

func withTraceContext(ctx context.Context, traceID, spanID string) context.Context {
	ctx = context.WithValue(ctx, traceIDKey, traceID)
	ctx = context.WithValue(ctx, spanIDKey, spanID)
	return ctx
}

func traceIDFromContext(ctx context.Context) string {
	if v, ok := ctx.Value(traceIDKey).(string); ok {
		return v
	}
	return ""
}

func spanIDFromContext(ctx context.Context) string {
	if v, ok := ctx.Value(spanIDKey).(string); ok {
		return v
	}
	return ""
}

func recordLatency(ms int64) {
	latencySum.Add(ms)
	latencyCount.Add(1)
	for i, bound := range bucketBounds {
		if float64(ms) <= bound {
			latencyBuckets[i].Add(1)
			return
		}
	}
	latencyBuckets[len(bucketBounds)].Add(1)
}

func recordRequest(statusCode int, latencyMs int64) {
	requestTotal.Add(1)
	if statusCode >= 400 {
		errorTotal.Add(1)
	}
	recordLatency(latencyMs)
}

func renderHistogram() string {
	var lines []string
	lines = append(lines, "# HELP http_request_duration_seconds Histogram of request durations")
	lines = append(lines, "# TYPE http_request_duration_seconds histogram")
	count := int64(0)
	for i, bound := range bucketBounds {
		count += latencyBuckets[i].Load()
		lines = append(lines, fmt.Sprintf(
			"http_request_duration_seconds_bucket{le=\"%g\"} %d",
			bound/1000.0, count,
		))
	}
	count += latencyBuckets[len(bucketBounds)].Load()
	lines = append(lines, fmt.Sprintf(
		"http_request_duration_seconds_bucket{le=\"+Inf\"} %d", count,
	))
	lines = append(lines, fmt.Sprintf(
		"http_request_duration_seconds_sum %f",
		float64(latencySum.Load())/1000.0,
	))
	lines = append(lines, fmt.Sprintf(
		"http_request_duration_seconds_count %d",
		latencyCount.Load(),
	))
	var result string
	for _, l := range lines {
		result += l + "\n"
	}
	return result
}

func renderCounters() string {
	var out string
	out += "# HELP http_requests_total Total HTTP requests\n"
	out += "# TYPE http_requests_total counter\n"
	out += fmt.Sprintf("http_requests_total %d\n", requestTotal.Load())
	out += "# HELP http_errors_total Total HTTP errors (status >= 400)\n"
	out += "# TYPE http_errors_total counter\n"
	out += fmt.Sprintf("http_errors_total %d\n", errorTotal.Load())
	return out
}

func metricsHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain; version=0.0.4")
	fmt.Fprint(w, renderCounters())
	fmt.Fprint(w, renderHistogram())
}

type observedHandler struct {
	next    http.HandlerFunc
	route   string
	service string
}

func (h *observedHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	traceID := r.Header.Get("traceparent")
	if traceID == "" {
		traceID = generateTraceID()
	}
	spanID := generateSpanID()
	ctx := withTraceContext(r.Context(), traceID, spanID)

	start := time.Now()

	rw := &responseWriter{ResponseWriter: w, statusCode: 200}
	h.next(rw, r.WithContext(ctx))

	latencyMs := time.Since(start).Milliseconds()
	recordRequest(rw.statusCode, latencyMs)

	level := "INFO"
	if rw.statusCode >= 500 {
		level = "ERROR"
	} else if rw.statusCode >= 400 {
		level = "WARN"
	}

	entry := LogEntry{
		Level:      level,
		Message:    "request completed",
		Service:    h.service,
		TraceID:    traceID,
		SpanID:     spanID,
		Route:      h.route,
		StatusCode: rw.statusCode,
		LatencyMs:  latencyMs,
	}
	if rw.statusCode >= 400 {
		entry.Error = fmt.Sprintf("HTTP %d on %s", rw.statusCode, h.route)
	}
	emitLog(entry)
}

type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

func healthHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

func usersHandler(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	userID := r.URL.Query().Get("id")
	if userID == "" {
		userID = strconv.Itoa(rand.Intn(1000) + 1)
	}

	childSpanID := generateSpanID()
	ctx = withTraceContext(ctx, traceIDFromContext(ctx), childSpanID)

	emitLog(LogEntry{
		Level:   "DEBUG",
		Message: "fetching user",
		Service: "user-service",
		TraceID: traceIDFromContext(ctx),
		SpanID:  childSpanID,
		UserID:  userID,
		Route:   "GET /users",
	})

	if rand.Float64() < 0.15 {
		time.Sleep(time.Duration(rand.Intn(2000)+500) * time.Millisecond)
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{
		"id":   userID,
		"name": fmt.Sprintf("user_%s", userID),
	})
}

func ordersHandler(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	orderID := r.URL.Query().Get("id")
	if orderID == "" {
		orderID = strconv.Itoa(rand.Intn(5000) + 1)
	}

	childSpanID := generateSpanID()
	ctx = withTraceContext(ctx, traceIDFromContext(ctx), childSpanID)

	emitLog(LogEntry{
		Level:   "DEBUG",
		Message: "processing order",
		Service: "order-service",
		TraceID: traceIDFromContext(ctx),
		SpanID:  childSpanID,
		Route:   "GET /orders",
	})

	if rand.Float64() < 0.1 {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusServiceUnavailable)
		json.NewEncoder(w).Encode(map[string]string{
			"error": "payment service unavailable",
		})
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{
		"order_id": orderID,
		"status":   "confirmed",
	})
}

func main() {
	mux := http.NewServeMux()
	mux.Handle("/health", &observedHandler{
		next: healthHandler, route: "GET /health", service: "api-server",
	})
	mux.Handle("/users", &observedHandler{
		next: usersHandler, route: "GET /users", service: "api-server",
	})
	mux.Handle("/orders", &observedHandler{
		next: ordersHandler, route: "GET /orders", service: "api-server",
	})
	mux.HandleFunc("/metrics", metricsHandler)

	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}

	emitLog(LogEntry{
		Level:   "INFO",
		Message: "server starting",
		Service: "api-server",
		Route:   fmt.Sprintf(":%s", port),
	})

	log.Fatal(http.ListenAndServe(":"+port, mux))
}